tls.Dial() validates hostname, no need to do that manually

ash2k opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

`tls.Dial()` validates hostname, no need to do that manually. `VerifyHostname()` is still needed for the `tls.Client()` code path. See https://github.com/kubernetes/kubernetes/pull/109750.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ash2k commented:
/sig api-machinery

ash2k commented:
/assign @liggitt

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/dial.go:
```diff
@@ -101,6 +100,21 @@ func dialURL(ctx context.Context, url *url.URL, transport http.RoundTripper) (ne
 				return nil, err
 			}
 
```
according to https://github.com/kubernetes/kubernetes/pull/109750, Handshake validates host... why is this still needed for this path?

liggitt replied:
also, should this switch to HandshakeContext (and tls.Dial below switch to tls.Dialer#DialContext?)

ash2k replied:
Follow up https://github.com/kubernetes/kubernetes/pull/110079

liggitt commented:
would like a second set of eyes on this as well
/assign @aojea

ash2k committed:
tls.Dial() validates hostname, no need to do that manually

Handshake() is still needed for tls.Client() code path. See https://github.com/kubernetes/kubernetes/pull/109750

aojea commented:
+1

roycaihw commented:
/triage accepted
